src/util/worker/mod.rs

```diff
@@ -275,22 +283,82 @@ where
     runner.shutdown();
 }
 
-// Fill buffer with next task batch comes from `rx`.
-fn fill_task_batch<T>(rx: &Receiver<Option<T>>, buffer: &mut Vec<T>, batch_size: usize) -> bool {
-    match rx.recv() {
-        Ok(Some(task)) => {
-            buffer.push(task);
-            while buffer.len() < batch_size {
-                match rx.try_recv() {
-                    Ok(Some(t)) => buffer.push(t),
-                    Err(TryRecvError::Empty) => break,
-                    Ok(None) | Err(_) => return false,
+fn poll_with_timer<R, T, U>(
+    mut runner: R,
+    rx: Receiver<Option<T>>,
+    counter: Arc<AtomicUsize>,
+    batch_size: usize,
+    mut timer: Timer<U>,
+) where
+    R: BatchRunnableWithTimer<T, U> + Send + 'static,
+    T: Display + Send + 'static,
+    U: Send + 'static,
+{
+    let name = thread::current().name().unwrap().to_owned();
+    let mut batch = Vec::with_capacity(batch_size);
+    let mut keep_going = true;
+    let mut tick_time = None;
+    while keep_going {
+        tick_time = tick_time.or_else(|| timer.next_timeout());
+        let timeout = tick_time.map(|t| t.checked_sub(Instant::now()));
+
+        keep_going = fill_task_batch(&rx, &mut batch, batch_size, timeout);
+
+        if !batch.is_empty() {
+            counter.fetch_sub(batch.len(), Ordering::SeqCst);
+            WORKER_PENDING_TASK_VEC
+                .with_label_values(&[&name])
+                .sub(batch.len() as f64);
+            WORKER_HANDLED_TASK_VEC
+                .with_label_values(&[&name])
+                .inc_by(batch.len() as f64)
+                .unwrap();
+            runner.run_batch(&mut batch);
+            batch.clear();
+        }
+
+        match tick_time {
+            Some(t) if t <= Instant::now() => {
```

Comment:
You need to store `Instant::now()` and use it for `pop_task_before`.

Reply:
what if `run_batch` takes too much time?

Reply:
That's the exact reason why `Instant::now` should be used rather than `t`.